flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -227,13 +241,19 @@ public void startNewWorker(ResourceProfile resourceProfile) {
 	}
 
 	@Override
-	public void stopWorker(ResourceID resourceID) {
-		// TODO: Implement to stop the worker
+	public void stopWorker(YarnWorkerNode workerNode) {
+		workerNodeMap.remove(workerNode.getResourceID().toString());
```

Comment:
@tillrohrmann I think the reason that we need the workerNodeMap is because that the task executor only has the resourceID when it tries to register with the resource manager. So the resource manager has to look up the YarnWorkerNode using the resouceID somehow when registerTaskExecutor is called. You can look at the workerStarted(ResourceID resourceID) method.

Reply:
I might have already asked you this, but looking at the code, I think we initialize the `ResourceID` on the `TaskManager` with `ContainerId.toString`. Can't we use this to reconstruct the `ContainerId` similar to http://hadoop.apache.org/docs/r2.8.0/api/org/apache/hadoop/yarn/api/records/ContainerId.html#fromString(java.lang.String)?

Reply:
This API is only available from hadoop 2.6. Dont we need to support 2.3 and 2.4 as well?

Reply:
Yes we do, if the application id did not change, then we could simply copy the code. But we have to add an attribution to it.

Reply:
Since we will need to keep workerNodeMap anyway, we can actually just look up the containerId using resourceId from workerNodeMap, there is no need to calculate the containerId from resourceId, so I think we dont need to use ContainerId.fromString. Also, copying the code will have compatibility issue if the fromString method differ in different hadoop version. What do you think?

Reply:
Yes this is correct. Let's keep the `workerNodeMap`. But then we have to make sure, that we always maintain a consistent view on the currently registered workers (e.g. properly removing entries upon failure, etc.)